pkg/controllers/user/alert/configsyncer/configsyncer.go

```diff
@@ -0,0 +1,259 @@
+
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	yaml "gopkg.in/yaml.v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+func NewConfigSyncer(ctx context.Context, cluster *config.UserContext, manager *manager.Manager) *ConfigSyncer {
+	return &ConfigSyncer{
+		secrets:            cluster.Core.Secrets("cattle-alerting"),
+		clusterAlertLister: cluster.Management.Management.ClusterAlerts(cluster.ClusterName).Controller().Lister(),
+		projectAlertLister: cluster.Management.Management.ProjectAlerts("").Controller().Lister(),
+		notifierLister:     cluster.Management.Management.Notifiers(cluster.ClusterName).Controller().Lister(),
+		clusterName:        cluster.ClusterName,
+		alertManager:       manager,
+	}
+}
+
+type ConfigSyncer struct {
+	secrets            v1.SecretInterface
+	projectAlertLister v3.ProjectAlertLister
+	clusterAlertLister v3.ClusterAlertLister
+	notifierLister     v3.NotifierLister
+	clusterName        string
+	alertManager       *manager.Manager
+}
+
+func (d *ConfigSyncer) ProjectSync(key string, alert *v3.ProjectAlert) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) ClusterSync(key string, alert *v3.ClusterAlert) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) NotifierSync(key string, alert *v3.Notifier) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) sync() error {
+	notifiers, err := d.notifierLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "List notifiers")
+	}
+
+	clusterAlerts, err := d.clusterAlertLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "List cluster alerts")
+	}
+
+	projectAlerts, err := d.projectAlertLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "List project alerts")
+	}
+
+	pAlerts := []*v3.ProjectAlert{}
+	for _, alert := range projectAlerts {
+		if controller.ObjectInCluster(d.clusterName, alert) {
+			pAlerts = append(pAlerts, alert)
+		}
+	}
+
+	config := d.alertManager.GetDefaultConfig()
+	config.Global.PagerdutyURL = "https://events.pagerduty.com/generic/2010-04-15/create_event.json"
+
+	d.addClusterAlert2Config(config, clusterAlerts, notifiers)
+	d.addProjectAlert2Config(config, pAlerts, notifiers)
+
+	data, err := yaml.Marshal(config)
+	if err != nil {
+		return errors.Wrapf(err, "Marshal secrets")
+	}
+
+	//TODO: check why lister does not work
+	configSecret, err := d.secrets.Get("alertmanager", metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "Get secrets")
+	}
+
+	if string(configSecret.Data["config.yml"]) != string(data) {
+		configSecret.Data["config.yml"] = data
+		configSecret.Data["email.tmpl"] = []byte(deploy.EmailTmlp)
+
+		_, err = d.secrets.Update(configSecret)
+		if err != nil {
+			return errors.Wrapf(err, "Update secrets")
+		}
+
+		go func() {
```

Comment:
I'm concerned with this go routine. Any reason why we can't call `d.alertManager.ReloadConfiguration()` directly? Also why do we do it 5 times in a loop, with delay, even when there was no failure?

Reply:
Ok, based on talk with @zionwu it looks like we do it in a loop to ensure that in the end the config map will get updated in the pod. It should be redone in a way that:

1) Controller only pushes the changes to config map
2) Alert manager microservice watches for config map updates, and calls the reload on itself.

Reply:
The change is done.  the micro-service is https://github.com/rancher/alertmanager-helper.